### DISCUSSION TITLE
NativeAOT and Trimming - Implications or best practices for organizing dependencies?

### DISCUSSION THREAD

**Discussion by csturm83** [dotnet/runtime#102534] [Q&A] [answered]
Let's say we have a library that for illustrative purposes we'll call 'LibABC'. We are considering breaking up the library into 3 distinct libs/NuGet packages since not all use cases will require all of the functionality provided by LibABC at once.

**Hypothetical Breakdown**
LibA = Core set of abstractions and base implementations of some features (for those who want to build only on top of LibA or don't need the features or implementations of LibB and LibC).
LibB => has dependency on abstractions or implementations in LibA
LibC => has dependency on abstractions or implementations in LibA

If LibABC is broken out into separate assemblies/packages as roughly outlined above, will there be an impact to NativeAOT/trimming capability or effectiveness if a user were to reference LibA, LibB, and LibC individually at the same time (say from a metapackage)?

In other words, is the compiler/linker's ability to trim impacted by breaking up an assembly into smaller assemblies for organizational purposes (E.g. for non-trimming scenarios where you only want to reference what you need)? Is it a matter of preference? Would there be any caveats or gotchas either way?

For the sake of simplicity in the example, let's say all code in LibABC is trimming compatible by default.

Thanks in advance for any advice or suggestions!

**Comment by huoyaoyuan** [comment]
Trimming are done for individual types and methods. I'm not aware that assembly boundary can affect anything. `Assembly.GetTypes` is surely impractical with trimming.

**Reply by MichalStrehovsky** [reply, marked as answer]
Assembly boundaries could matter with `<TrimMode>partial</TrimMode>`. In this mode, that is not preferred, but some workloads like Blazor WASM/MAUI default to it for historical reasons, any assembly that wasn't marked `<IsTrimmable>true</IsTrimmable>` in the library project file would be kept unconditionally. There could be differences in TrimMode=partial if _some_ of the new assemblies get marked IsTrimmable, but not others. Otherwise it's all the same.
